mgmtfn/dockplugin/netDriver.go

```diff
@@ -523,3 +630,171 @@ func GetDockerNetworkName(nwID string) (string, string, string, error) {
+	epgList, err := epgCfg.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+	var epg *mastercfg.EndpointGroupState
+	found := false
+	for _, epgP := range epgList {
+		epg = epgP.(*mastercfg.EndpointGroupState)
+		if epg.GroupTag == epgTag {
+			found = true
+			break
+		}
+	}
+	if !found {
+		err = errors.New("Couldn't find group matching the tag")
+		return nil, err
+	}
+	return epg, nil
+}
+
+// FindNetworkFromTag finds the network that has matching tag
+func FindNetworkFromTag(nwTag string) (*mastercfg.CfgNetworkState, error) {
+	stateDriver, err := utils.GetStateDriver()
+	if err != nil {
+		return nil, err
+	}
+	nwCfg := &mastercfg.CfgNetworkState{}
+	nwCfg.StateDriver = stateDriver
+	nwList, err := nwCfg.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+	var nw *mastercfg.CfgNetworkState
+	found := false
+	for _, nwP := range nwList {
+		nw = nwP.(*mastercfg.CfgNetworkState)
+		if nw.NetworkTag == nwTag {
+			found = true
+			break
+		}
+	}
+	if !found {
+		err = errors.New("Couldn't find network matching the tag")
+		return nil, err
+	}
+	return nw, nil
+}
+
+// createNetworkHelper creates the association between docker network and contiv network
+//  if tag is given map docker net to epg or network, else create a contiv network
+func createNetworkHelper(networkID string, tag string, IPv4Data, IPv6Data []driverapi.IPAMData) error {
+	var tenantName, networkName, serviceName string
+	var err error
+	if tag != "" {
+		// we need to map docker network to policy group or network using the tag
+		log.Infof("Received tag %s", tag)
+		var nw *mastercfg.CfgNetworkState
+		epg, err := FindGroupFromTag(tag)
+		if err != nil {
+			nw, err = FindNetworkFromTag(tag)
+			if err != nil {
+				return errors.New("failed to lookup tag")
+			}
+		}
+		if epg != nil {
+			tenantName = epg.TenantName
+			networkName = epg.NetworkName
+			serviceName = epg.GroupName
+		} else if nw != nil {
+			tenantName = nw.Tenant
+			networkName = nw.NetworkName
+			serviceName = ""
+		}
+	} else if len(IPv4Data) > 0 {
+		// if subnet is specified in docker command, we create a contiv network
+		subnetPool := ""
+		gateway := ""
+		if IPv4Data[0].Pool != nil {
+			subnetPool = IPv4Data[0].Pool.String()
+		}
+		if IPv4Data[0].Gateway != nil {
+			gateway = strings.Split(IPv4Data[0].Gateway.String(), "/")[0]
+		}
+		subnetv6 := ""
+		gatewayv6 := ""
+		if len(IPv6Data) > 0 {
+			if IPv6Data[0].Pool != nil {
+				subnetv6 = IPv6Data[0].Pool.String()
+			}
+			if IPv6Data[0].Gateway != nil {
+				gatewayv6 = strings.Split(IPv6Data[0].Gateway.String(), "/")[0]
+			}
+		}
+		// build key and URL
+		keyStr := "default" + ":" + networkID
+		url := "/api/v1/networks/" + keyStr + "/"
+
+		tenantName = "default"
+		networkName = networkID
+		serviceName = ""
+
+		req := client.Network{
+			TenantName:  tenantName,
+			NetworkName: networkName,
+			Subnet:      subnetPool,
+			Gateway:     gateway,
+			Ipv6Subnet:  subnetv6,
+			Ipv6Gateway: gatewayv6,
+			Encap:       "vlan",
```

Comment:
When user creates a docker network without a tag, we need to create a contiv network. Since we don't know the encap to use, we need to pick a default. Because of the issues in vxlan, I picked vlan encap. Now that vxlan issues are fixed by changing the UDP port, we could change the default encap to vxlan if it is required.